### PR TITLE
CSHARP-2804: Support for JavaScript User Defined Functions.

### PR DESCRIPTION
**Pull request by DmitryLukyanov**
EG: https://evergreen.mongodb.com/version/5e8f569a0ae60673bc4d40dc
I confirmed that the below expressions are not the target of this ticket:
* $_internalJs
* $_internalJsReduce
* $where

The target is adding a test for aggregate `$function.`

### CODE REVIEW

**Comment by DmitryLukyanov** [comment]
Waiting for EG

**Comment by DmitryLukyanov** [comment]
ready for review

File: tests/MongoDB.Driver.Tests/AggregateFluentTests.cs [hunk 1]
```diff
@@ -196,6 +196,71 @@ public void ChangeStream_should_add_the_expected_stage_when_options_is_null()
...
+            var databaseName = "test";
+            var collectionName = "collection";
+            DropCollection(client, databaseName, collectionName);
+            var collection = client.GetDatabase(databaseName).GetCollection<BsonDocument>(collectionName);
+            var players = new[]
+            {
+                BsonDocument.Parse("{ _id : 1, name : 'Miss Cheevous',  scores : [10, 5, 10] }"),
+                BsonDocument.Parse("{ _id : 2, name : 'Miss Ann Thrope', scores : [10, 10, 10] }"),
+                BsonDocument.Parse("{ _id : 3, name : 'Mrs. Eppie Delta', scores : [9, 8, 8] }")
+            };
```
Comment (rstam) [inline tests/MongoDB.Driver.Tests/AggregateFluentTests.cs:214]: Nit: use spaces after opening braces and before closing braces (including array braces):

    var players = new[]
    {
        BsonDocument.Parse("{ _id : 1, name : 'Miss Cheevous',  scores : [ 10, 5, 10 ] }"),
        BsonDocument.Parse("{ _id : 2, name : 'Miss Ann Thrope', scores : [ 10, 10, 10 ] }"),
        BsonDocument.Parse("{ _id : 3, name : 'Mrs. Eppie Delta', scores : [ 9, 8, 8 ] }")
    };

File: tests/MongoDB.Driver.Tests/AggregateFluentTests.cs [hunk 2]
```diff
@@ -196,6 +196,71 @@ public void ChangeStream_should_add_the_expected_stage_when_options_is_null()
...
+            collection.InsertMany(players);
+
+            var isFoundBodyFunction = "function(name) { return hex_md5(name) == '15b0a220baa16331e8d80e15367677ad' }";
```
Comment (rstam) [inline tests/MongoDB.Driver.Tests/AggregateFluentTests.cs]: Rename to: `isFoundFunction` (rename `isFoundFunction` below first)

File: tests/MongoDB.Driver.Tests/AggregateFluentTests.cs [hunk 3]
```diff
@@ -196,6 +196,71 @@ public void ChangeStream_should_add_the_expected_stage_when_options_is_null()
...
+            var messageBodyFunction = "function(name, scores) { let total = Array.sum(scores); return `Hello ${ name}. Your total score is ${ total}.` }";
```
Comment (rstam) [inline tests/MongoDB.Driver.Tests/AggregateFluentTests.cs]: Rename to: `messageFunction` (rename `isFoundFunction` below first)

File: tests/MongoDB.Driver.Tests/AggregateFluentTests.cs [hunk 4]
```diff
@@ -196,6 +196,71 @@ public void ChangeStream_should_add_the_expected_stage_when_options_is_null()
...
+            var isFoundFunction = new BsonDocument
```
Comment (rstam) [inline tests/MongoDB.Driver.Tests/AggregateFluentTests.cs]: Rename to: `isFoundExpression`

File: tests/MongoDB.Driver.Tests/AggregateFluentTests.cs [hunk 5]
```diff
@@ -196,6 +196,71 @@ public void ChangeStream_should_add_the_expected_stage_when_options_is_null()
...
+                    "$function",
+                    new BsonDocument
+                    {
+                        { "body", isFoundBodyFunction },
+                        { "args", new BsonArray { "$name" } },
+                        { "lang", "js" }
+                    }
+                },
+            };
+            var messageFunction = new BsonDocument
```
Comment (rstam) [inline tests/MongoDB.Driver.Tests/AggregateFluentTests.cs]: Rename to: `messageExpression`

File: tests/MongoDB.Driver.Tests/AggregateFluentTests.cs [hunk 6]
```diff
@@ -196,6 +196,71 @@ public void ChangeStream_should_add_the_expected_stage_when_options_is_null()
...
+                    "$function",
+                    new BsonDocument
+                    {
+                        { "body", messageBodyFunction },
+                        { "args", new BsonArray { "$name", "$scores" } },
+                        { "lang", "js" }
+                    }
+                },
+            };
+            var aggregatePipelineStage = new BsonDocument
```
Comment (rstam) [inline tests/MongoDB.Driver.Tests/AggregateFluentTests.cs]: Rename to: `addFieldsStage`

File: tests/MongoDB.Driver.Tests/AggregateFluentTests.cs [hunk 7]
```diff
@@ -196,6 +196,71 @@ public void ChangeStream_should_add_the_expected_stage_when_options_is_null()
...
+                {
+                    "$addFields",
+                    new BsonDocument
+                    {
+                        { "isFound", isFoundFunction },
+                        { "message", messageFunction },
+                    }
+                }
+            };
+            var aggregateStagePipelineDefinition = new BsonDocumentStagePipelineDefinition<BsonDocument, BsonDocument>(new[] { aggregatePipelineStage });
```
Comment (rstam) [inline tests/MongoDB.Driver.Tests/AggregateFluentTests.cs]: Replace 254 with:

    var stages = new[] { addFieldsStage };
    var pipeline = new BsonDocumentStagePipelineDefinition<BsonDocument, BsonDocument>(stages);

File: tests/MongoDB.Driver.Tests/AggregateFluentTests.cs [hunk 8]
```diff
@@ -196,6 +196,71 @@ public void ChangeStream_should_add_the_expected_stage_when_options_is_null()
...
+
+            var result = collection.Aggregate(aggregateStagePipelineDefinition).ToList();
+
+            result.Count.Should().Be(3);
+            result[0].Should().Be("{ _id : 1, name : 'Miss Cheevous', scores : [10, 5, 10], isFound : false, message : 'Hello Miss Cheevous. Your total score is 25.' }");
+            result[1].Should().Be("{ _id : 2, name : 'Miss Ann Thrope', scores : [10, 10, 10], isFound : true, message : 'Hello Miss Ann Thrope. Your total score is 30.' }");
+            result[2].Should().Be("{ _id : 3, name : 'Mrs. Eppie Delta', scores : [9, 8, 8], isFound : false, message : 'Hello Mrs. Eppie Delta. Your total score is 25.' }");
```
Comment (rstam) [inline tests/MongoDB.Driver.Tests/AggregateFluentTests.cs]: nit: use spaces after opening brace and before closing brace:

    result[0].Should().Be("{ _id : 1, name : 'Miss Cheevous', scores : [ 10, 5, 10 ], isFound : false, message : 'Hello Miss Cheevous. Your total score is 25.' }");
    result[1].Should().Be("{ _id : 2, name : 'Miss Ann Thrope', scores : [ 10, 10, 10 ], isFound : true, message : 'Hello Miss Ann Thrope. Your total score is 30.' }");
    result[2].Should().Be("{ _id : 3, name : 'Mrs. Eppie Delta', scores : [ 9, 8, 8 ], isFound : false, message : 'Hello Mrs. Eppie Delta. Your total score is 25.' }");

File: tests/MongoDB.Driver.Tests/AggregateFluentTests.cs (hunk 1 above)
Comment (rstam) [reply tests/MongoDB.Driver.Tests/AggregateFluentTests.cs:214]: Note; I've added this convention to our evolving style guide

**Review by rstam — CHANGES_REQUESTED** [review]
The new test is good.

I'm requesting some name changes and some style changes.

At some point we're going to point users to this example so it's important that the names be as accurate and self-explanatory as possible.

File: tests/MongoDB.Driver.Tests/AggregateFluentTests.cs (hunk 1 above)
Comment (DmitryLukyanov) [reply tests/MongoDB.Driver.Tests/AggregateFluentTests.cs:214]: As I remember, in the previous similar dispute we decided that array braces should be without opening/ending spaces :) But ok, done

File: tests/MongoDB.Driver.Tests/AggregateFluentTests.cs (hunk 2 above)
Comment (DmitryLukyanov) [reply tests/MongoDB.Driver.Tests/AggregateFluentTests.cs]: done

File: tests/MongoDB.Driver.Tests/AggregateFluentTests.cs (hunk 3 above)
Comment (DmitryLukyanov) [reply tests/MongoDB.Driver.Tests/AggregateFluentTests.cs]: done

File: tests/MongoDB.Driver.Tests/AggregateFluentTests.cs (hunk 4 above)
Comment (DmitryLukyanov) [reply tests/MongoDB.Driver.Tests/AggregateFluentTests.cs]: done

File: tests/MongoDB.Driver.Tests/AggregateFluentTests.cs (hunk 5 above)
Comment (DmitryLukyanov) [reply tests/MongoDB.Driver.Tests/AggregateFluentTests.cs]: done

File: tests/MongoDB.Driver.Tests/AggregateFluentTests.cs (hunk 6 above)
Comment (DmitryLukyanov) [reply tests/MongoDB.Driver.Tests/AggregateFluentTests.cs]: done

File: tests/MongoDB.Driver.Tests/AggregateFluentTests.cs (hunk 7 above)
Comment (DmitryLukyanov) [reply tests/MongoDB.Driver.Tests/AggregateFluentTests.cs]: done

File: tests/MongoDB.Driver.Tests/AggregateFluentTests.cs (hunk 8 above)
Comment (DmitryLukyanov) [reply tests/MongoDB.Driver.Tests/AggregateFluentTests.cs]: done

**Comment by DmitryLukyanov** [comment]
ready for review

File: tests/MongoDB.Driver.Tests/AggregateFluentTests.cs (hunk 1 above)
Comment (rstam) [reply tests/MongoDB.Driver.Tests/AggregateFluentTests.cs:214]: I searched our code base and we are 50/50 on this one.

**Review by rstam — APPROVED** [review]
LGTM

**Pull request closed without merging**